File: другое/тинькофф.аисд/1/J.ts
const n = 5;
const arr: [number, number][] = [
  [1, 3],
  [4, 7],
  [1, 10],
  [3, 3],
  [3, 5],
  [3, 1],
  [10, 1],
];

for (const item of arr) {
  console.log("Item", item);
  const [n, m] = item;
  if (n === 1) {
    let best = 0;
    let div = Infinity;
    const S = (m * (m + 1)) / 2;
    const avg = Math.floor(S / 2);
    for (let i = m - 1; i >= 1; i--) {
      const b = (i * (i + 1)) / 2;
      const db = Math.abs(avg - b);
      if (db < div) {
        div = db;
        best = i;
      } else {
        break;
      }
    }
    console.log("V " + String(best + 1));
  } else if (m === 1) {
    let best = 0;
    let div = Infinity;
    const S = (n * (n + 1)) / 2;
    const avg = Math.floor(S / 2);
    for (let i = n - 1; i >= 1; i--) {
      const b = (i * (i + 1)) / 2;
      const db = Math.abs(avg - b);
      if (db < div) {
        div = db;
        best = i;
      } else {
        break;
      }
    }
    console.log("H " + String(best + 1));
  } else {
    const S = (m * n * (m * n + 1)) / 2;
    let sigma = 0;
    for (let i = 0; i < n; i++) {
      sigma += 1 + m * i;
    }
    const x = Math.floor((0.5 * S + n) / (sigma + n));
    const valueX = x * (sigma + n) - n;
    const divX = S - 2 * valueX;
    const ansY = solveQuadraticEquation(m ** 2, m, -S);
    if (ansY === null) {
      console.log("V " + String(x + 1));
    } else if (ansY.length === 1) {
      const y = ansY[0];
      const valueY = Math.floor((y * m * (y * m + 1)) / 2);
      const divY = S - 2 * valueY;
      if (divX < divY) {
        console.log("V " + String(x + 1));
      } else {
        console.log("H " + String(y + 1));
      }
    }
  }
}

function solveQuadraticEquation(
  A: number,
  B: number,
  C: number,
): number[] | null {
  const discriminant = B * B - 4 * A * C;
  if (discriminant < 0) {
    return null;
  } else {
    const x1 = Math.floor((-B + Math.sqrt(discriminant)) / (2 * A));
    const x2 = Math.floor((-B - Math.sqrt(discriminant)) / (2 * A));
    if (x1 <= 0 && x2 <= 0) {
      return null;
    } else if (x1 <= 0) {
      return [x2];
    } else if (x2 <= 0) {
      return [x1];
    } else {
      return [x1, x2];
    }
  }
}

File: другое/тинькофф.аисд/1/J.py
import math

ln = int(input())

arr = []

for _ in range(ln):
    line = input().strip()
    arr.append(list(map(int, line.split())))


def L(n, m, k):
    return (((n - 1) * m * n + (k + 1) * n) * k) / 2


def R(n, m, k):
    return (((n - 1) * m * n + (k + 1 + m) * n) * (m - k)) / 2


def U(n, m, k):
    return ((k * m ** 2 + m) * k) / 2


def D(n, m, k):
    return (((k + n) * m ** 2 + m) * (n - k)) / 2


def horizontal(n, m):
    def binary():
        low = 1
        high = n
        while low <= high:
            mid = (low + high) // 2
            if U(n, m, mid) <= D(n, m, mid):
                low = mid + 1
            else:
                high = mid - 1
        return low - 1

    roots = findRoots(2 * m ** 2, 2 * m, -(n ** 2 * m ** 2) - n * m)
    if len(roots) == 1 and roots[0] < n:
        k = int(roots[0])
        div1 = abs(D(n, m, k) - U(n, m, k))
        div2 = abs(U(n, m, k + 1) - D(n, m, k + 1))
        if div1 < div2:
            return [k, div1]
        else:
            return [k + 1, div2]
    else:
        k = binary()
        div1 = abs(D(n, m, k) - U(n, m, k))
        div2 = abs(U(n, m, k + 1) - D(n, m, k + 1))
        if div1 < div2:
            return [k, div1]
        else:
            return [k + 1, div2]


def vertical(n, m):
    def binary():
        low = 1
        high = m
        while low <= high:
            mid = (low + high) // 2
            if L(n, m, mid) <= R(n, m, mid):
                low = mid + 1
            else:
                high = mid - 1
        return low - 1

    roots = findRoots(-2 * n, 2 * m * n - 2 * m * n ** 2 - 2 * n, m ** 2 * n ** 2 + m * n)
    if len(roots) == 1 and roots[0] < m:
        k = int(roots[0])
        div1 = abs(L(n, m, k) - R(n, m, k))
        div2 = abs(L(n, m, k + 1) - R(n, m, k + 1))
        if div1 < div2:
            return [k, div1]
        else:
            return [k + 1, div2]
    else:
        k = binary()
        div1 = abs(L(n, m, k) - R(n, m, k))
        div2 = abs(L(n, m, k + 1) - R(n, m, k + 1))
        if div1 < div2:
            return [k, div1]
        else:
            return [k + 1, div2]


def findRoots(A, B, C):
    discriminant = B ** 2 - 4 * A * C
    if discriminant > 0:
        root1 = (-B + math.sqrt(discriminant)) / (2 * A)
        root2 = (-B - math.sqrt(discriminant)) / (2 * A)
        if root1 < 0 and root2 < 0:
            return []
        elif root1 < 0:
            return [root2]
        elif root2 < 0:
            return [root1]
        else:
            return [root1, root2]
    elif discriminant == 0:
        root = -B / (2 * A)
        if root < 0:
            return []
        else:
            return [root]
    else:
        return []


for item in arr:
    n, m = item
    vert = vertical(n, m)
    hor = horizontal(n, m)
    if vert[1] < hor[1]:
        print("V", vert[0] + 1)
    else:
        print("H", hor[0] + 1)
